invalidate_request.py

```diff
@@ -0,0 +1,417 @@
+"""
+This module allows to invalidate a 
+root requests and its generated chains.
+This is intended to be used only for
+`production_manager` users or `administrators`.
+"""
+
+# Add path to code that allow easy access to McM
+# sys.path.append("/afs/cern.ch/cms/PPD/PdmV/tools/McM/")
+
+import datetime
+import logging
+import pprint
+import sys
+from itertools import groupby
+from copy import deepcopy
+from rest import McM
+
+# Check if version is >= 3.11, otherwise raise an exception.
+if sys.version_info < (3, 11):
+    msg = (
+        "Unsuported Python version.\n"
+        "Please use a version equal of higher than Python 3.11.\n"
+        "Current version: %s " % (sys.version)
+    )
+    raise RuntimeError(msg)
+
+
+class InvalidateDeleteRequests:
+    """
+    Invalidate and delete all the chain requests in McM
+    linked to a root request in McM and delete the
+    root requests if required.
+    """
+
+    def __init__(self, mcm: McM) -> None:
+        self.mcm = mcm
+        self.logger = self._get_logger()
+        self._chain_request_type = "chained_requests"
+        self._request_type = "requests"
+        self.logger.warning("Sending requests to target environment: %s", self.mcm.host)
+
+    def _get_logger(self) -> logging.Logger:
+        """
+        Creates a logger to record the performed steps.
+        """
+        logger: logging.Logger = logging.getLogger(__name__)
+        formatter = logging.Formatter("[%(asctime)s][%(levelname)s] %(message)s")
+        current_date = str(datetime.datetime.now().strftime("%Y_%m_%d_%H_%M"))
+        fh: logging.Handler = logging.FileHandler(
+            f"mcm_invalidate_requests_{current_date}.log"
+        )
+        fh.setFormatter(formatter)
+        logger.addHandler(fh)
+        return logger
+
+    def _pretty(self, obj: dict) -> str:
+        """
+        Pretty print an object in console
+        """
+        return pprint.pformat(obj, width=50, compact=True)
+
+    def _get(self, endpoint: str) -> dict:
+        """
+        Execute a raw GET operation for a resource in McM.
+        """
+        return self.mcm._McM__get(endpoint)
+
+    def _get_invalidations(self, requests: list[str]) -> list[dict]:
+        """
+        Get all the invalidation records related to request given by
+        parameter.
+
+        Args:
+            requests (list[str]): List of `prepids` used to retrieve its invalidations.
+
+        Returns:
+            list[str]: Invalidations related to the requests.
+        """
+        results: list[dict] = []
+        for req in requests:
+            inv_req = self.mcm.get("invalidations", query=f"prepid={req}")
```

Comment:
Do we have a way to perform bulk requests?

Reply:
We could perform a wildcard query to retrieve them with a single call but this would require more code to pick the appropriate prefix. If it is related to something else, please could you provide more details?

Reply:
I was thinking of something like `invalidations?prepid=a,b,c,d` that's sometimes supported in McM, but it seems this is unfortunately not supported.